### PR TITLE
Use SWR instead of rolling our own api fetcher hooks

### PR DESCRIPTION
**Pull request by alanwguo**
## Why are these changes needed?
SWR is fully-featured and supports auto-refresh and other features. THis also fixes a bug where the fetchers were not cleaned up properly and it was possible to continue doing fetching for pages a user was previously on.

Signed-off-by: Alan Guo <aguo@anyscale.com>

## Checks
Tested rapidly switching between multiple tabs in the UI. Fetchers were properly cleaned up.
Checked every single UI page to make sure they continue to load.

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/client/src/pages/job/JobTaskNameProgressTable/JobTaskNameProgressTable.tsx [hunk 1]
```diff
@@ -72,7 +72,7 @@ export const JobTaskNameProgressTable = ({
           {progress.map(
             ({ name, progress, numFailed, numActive, numFinished }) => {
               return (
-                <TableRow>
+                <TableRow key={name}>
```
Comment (alanwguo) [inline dashboard/client/src/pages/job/JobTaskNameProgressTable/JobTaskNameProgressTable.tsx:75]: THis is orthogonal change. But it cleans up a warning in the console.

**Review by rkooo567 — COMMENTED** [review]
Do we need tests? What's the best way to test this change? Should we do some QA after merging this?

Also a couple config questions. https://swr.vercel.app/docs/options
- shouldRetryOnError -> should we disable this? Do you know what "error" means here? Does it include status like 500? 
- Is it possible to make it not send another request when the previous request is not replied? Should we use `onLoadingSlow` to  disable the next polling or sth?

**Comment by alanwguo** [comment]
> Do we need tests? What's the best way to test this change? Should we do some QA after merging this?
> 
> Also a couple config questions. https://swr.vercel.app/docs/options
> 
> * shouldRetryOnError -> should we disable this? Do you know what "error" means here? Does it include status like 500?
> * Is it possible to make it not send another request when the previous request is not replied? Should we use `onLoadingSlow` to  disable the next polling or sth?

We disable refreshing if we encounter an error. ShouldRetryOnError refers to just retrying once. I think this behavior is fine.
This already waits for the request to finish before attempting a second request.

Testing-wise, I tested every endpoint that was modified. For automated tests, we can do this once we start adding tests to each page.

**Comment by rkooo567** [comment]
sgtm. Let's merge it and see if this fixes the issue in the master!

**Review by rkooo567 — APPROVED** [review]

File: dashboard/client/src/pages/job/hook/useJobDetail.ts [hunk 2]
```diff
@@ -1,48 +1,29 @@
...
+        const rsp = await getJobDetail(params.id);
+        return rsp.data;
+      } catch (e) {
+        setMsg("Job Query Error Please Check JobId");
+        setRefresh(false);
       }
-    };
-  }, [getJob]);
+    },
+    { refreshInterval: refreshing ? 4000 : 0 },
```
Comment (galenhwang) [inline dashboard/client/src/pages/job/hook/useJobDetail.ts:25]: nit: extract 4000 to constant

**Review by galenhwang — APPROVED** [review]

**Comment by rkooo567** [comment]
I am going to verify in the master if this fixes the issue tomorrow!
